tests/server/test_server.py

```diff
@@ -0,0 +1,536 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+
+import os
+import sys
+import base64
+import json
+
+from mock import patch, Mock
+
+from tank_test.tank_test_base import setUpModule # noqa
+
+import sgtk
+from tank_vendor.shotgun_api3.lib.mockgun import Shotgun
+
+repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
+fixtures_root = os.path.join(repo_root, "tests", "fixtures")
+
+sys.path.insert(0, os.path.join(repo_root, "python"))
+
+# Mock Qt since we don't have it.
+sgtk.platform.qt.QtCore = Mock()
+sgtk.platform.qt.QtGui = Mock()
+
+# Doing this import will add the twisted librairies
+import tk_framework_desktopserver # noqa
+
+from twisted.trial import unittest
+from twisted.internet import ssl
+from autobahn.twisted.websocket import connectWS, WebSocketClientFactory, WebSocketClientProtocol
+from twisted.internet.defer import Deferred
+from twisted.internet import reactor
+from cryptography.fernet import Fernet
+
+from twisted.internet import base
+base.DelayedCall.debug = True
+
+
+class MockShotgunApi(object):
+    """
+    Mocks the v2 protocol with a custom method.
+    """
+    PUBLIC_API_METHODS = ["repeat_value", "missing_attribute"]
+
+    def __init__(self, host, process_manager, wss_key):
+        self._host = host
+
+    def repeat_value(self, payload):
+        self._host.reply({"value": payload["value"] * 3})
+
+
+def TestServerBase(class_name, class_parents, class_attr):
```

Comment:
Seems like twisted.trial has issue with unit test base classes, so we're using a metaclass instead to add basic functionality to the test classes.